File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/115-distinct-subsequences-approach3.go
package main

import "fmt"

func numDistinct(s string, t string) int {
	m, n := len(s), len(t)
	dp := make([]int, n+1)
	dp[0] = 1
	for i := 0; i < m; i++ {
		for j := n; j >= 1; j-- {
			if s[i] == t[j-1] {
				dp[j] += dp[j-1]
			}
		}
	}
	return dp[n]
}

func main() {
	assert := func(cond bool) {
		if !cond {
			panic("assertion failed")
		}
	}
	assert(numDistinct("rabbbit", "rabbit") == 3)
	assert(numDistinct("babgbag", "bag") == 5)
	assert(numDistinct("abc", "") == 1)
	assert(numDistinct("", "a") == 0)
	assert(numDistinct("abc", "abc") == 1)
	assert(numDistinct("aaa", "b") == 0)
	fmt.Println("all tests pass")
}

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/115-distinct-subsequences-approach3.py
def num_distinct(s: str, t: str) -> int:
    m, n = len(s), len(t)
    dp = [0] * (n + 1)
    dp[0] = 1
    for i in range(m):
        for j in range(n, 0, -1):
            if s[i] == t[j - 1]:
                dp[j] += dp[j - 1]
    return dp[n]

assert num_distinct("rabbbit", "rabbit") == 3
assert num_distinct("babgbag", "bag") == 5
assert num_distinct("abc", "") == 1
assert num_distinct("", "a") == 0
assert num_distinct("abc", "abc") == 1
assert num_distinct("aaa", "b") == 0
print("all tests pass")
